### PR TITLE
planner: refactor `getFullAnalyzeColumnsInfo`

### PR DESCRIPTION
**Pull request by hi-rustin**
### What problem does this PR solve?

Issue Number: None
Problem Summary:

### What is changed and how it works?

- refactored `getFullAnalyzeColumnsInfo`
- Added tests for `handleAnalyzeOptionsV2`

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/planbuilder.go [hunk 1]
```diff
@@ -2356,45 +2348,86 @@ func (b *PlanBuilder) getFullAnalyzeColumnsInfo(
...
+// Helper function to get a list of column infos from a set of column IDs.
+func getColumnListFromSet(columns []*model.ColumnInfo, colSet map[int64]struct{}) []*model.ColumnInfo {
+	colList := make([]*model.ColumnInfo, 0, len(colSet))
+	for _, colInfo := range columns {
+		if _, ok := colSet[colInfo.ID]; ok {
+			colList = append(colList, colInfo)
+		}
+	}
+	return colList
+}
```
Comment (hawkingrei) [inline planner/core/planbuilder.go:2429]: We can use genric to merge three functions into one.

**Review by hawkingrei — APPROVED** [review]

File: planner/core/planbuilder.go [hunk 2]
```diff
@@ -3009,6 +3053,10 @@ func parseAnalyzeOptionsV2(opts []ast.AnalyzeOpt) (map[ast.AnalyzeOptionType]uin
 	if sampleNum > 0 && sampleRate > 0 {
 		return nil, errors.Errorf("You can only either set the value of the sample num or set the value of the sample rate. Don't set both of them")
 	}
+
+	if optMap[ast.AnalyzeOptCMSketchWidth]*optMap[ast.AnalyzeOptCMSketchDepth] > CMSketchSizeLimit {
+		return nil, errors.Errorf("cm sketch size(depth * width) should not larger than %d", CMSketchSizeLimit)
+	}
```
Comment (time-and-fate) [inline planner/core/planbuilder.go:3059]: Stats v2 doesn't have cm sketch anymore, so this is no longer a necessary check.
I think we don't need this error, or tell the user that this option is useless in the error message.

File: planner/core/planbuilder.go [hunk 3]
```diff
@@ -2438,30 +2464,36 @@ func getModifiedIndexesInfoForAnalyze(sctx sessionctx.Context, tblInfo *model.Ta
...
 	physicalIDs []int64,
-	names []string,
+	partitionNames []string,
 	tbl *ast.TableName,
 	version int,
 	persistOpts bool,
 	rsOptionsMap map[int64]V2AnalyzeOptions,
 ) ([]AnalyzeColumnsTask, error) {
+	// Version 2 doesn't support incremental analyze.
+	// And incremental analyze will deprecated in the future.
```
Comment (time-and-fate) [inline planner/core/planbuilder.go:2476]:
```suggestion
	// And incremental analyze will be deprecated in the future.
```

File: planner/core/planbuilder.go (hunk 2 above)
Comment (hi-rustin) [reply planner/core/planbuilder.go]: Addressed. See https://github.com/pingcap/tidb/pull/46044/commits/3c399bfe47411e914eae369dfc711da5bfb13029

I removed that check and also checked the version first in handleAnalyzeOptions.

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**